File: packages/alteration/src/main/java/fdit/alteration/core/basestation/BaseStationParser.java
package fdit.alteration.core.basestation;

import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonParser;
import fdit.alteration.core.basestation.message.BaseStationMessage.BstMessageTypeSwitch;
import fdit.alteration.core.basestation.message.BaseStationMessageFull;
import fdit.alteration.core.engine.Message;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Optional;
import java.util.concurrent.Callable;

import static java.lang.Double.parseDouble;
import static java.lang.Integer.parseInt;
import static java.util.Arrays.copyOf;
import static java.util.Optional.empty;
import static java.util.Optional.of;
import static java.util.TimeZone.getTimeZone;

public class BaseStationParser {

    public static final String BST_DATE_PATTERN = "yyyy/MM/dd,HH:mm:ss.SSS";
    private static final String SPLIT_PATTERN = ",(?=(?:[^\"]*\"[^\"]*\")*[^\"]*$)";
    private static final String SPLIT_JSON_PATTERN = ",\\{";

    private BaseStationParser() {
    }

    public static Optional<Message> createBstMessage(final String line) {
        /* Separation du message de base et de l'objet json **/
        String[] parts = line.split(SPLIT_JSON_PATTERN);
        final boolean extraFieldsArePresent = parts.length > 1;

        /* Separation des valeurs du message **/
        String[] fields = parts[0].split(SPLIT_PATTERN);

        /* Regroupement des valeurs et de l'objet json **/
        final String[] fieldsWithElement = copyOf(fields, extraFieldsArePresent ? fields.length + 1 : fields.length);
        if (extraFieldsArePresent) fieldsWithElement[fieldsWithElement.length - 1] = '{' + parts[1];

        try {
            final int bstType = extractBstType(fieldsWithElement);
            final int sessionID = extractSessionID(fieldsWithElement);
            final int aircraftID = extractAircraftID(fieldsWithElement);
            final String icao = extractIcao(fieldsWithElement);
            final int flightID = extractFlightID(fieldsWithElement);
            final long timestampGenerated = extractTimestampGenerated(fieldsWithElement);
            final long timestampLogged = extractTimestampLogged(fieldsWithElement);

            return new BstMessageTypeSwitch<Optional<Message>>() {

                @Override
                public Optional<Message> visitBstMessageFull() {
                    return of(new BaseStationMessageFull(
                            bstType,
                            sessionID,
                            aircraftID,
                            icao,
                            flightID,
                            timestampGenerated,
                            timestampLogged,
                            nullableString(() -> extractCallSign(fieldsWithElement)),
                            nullableInteger(() -> extractAltitude(fieldsWithElement)),
                            nullableDouble(() -> extractGroundSpeed(fieldsWithElement)),
                            nullableDouble(() -> extractTrack(fieldsWithElement)),
                            nullableDouble(() -> extractLatitude(fieldsWithElement)),
                            nullableDouble(() -> extractLongitude(fieldsWithElement)),
                            nullableInteger(() -> extractVerticalRate(fieldsWithElement)),
                            nullableInteger(() -> extractSquawk(fieldsWithElement)),
                            nullableBoolean(() -> extractAlert(fieldsWithElement)),
                            nullableBoolean(() -> extractEmergency(fieldsWithElement)),
                            nullableBoolean(() -> extractSpi(fieldsWithElement)),
                            nullableBoolean(() -> extractOnGround(fieldsWithElement)),
                            nullableJsonObject(() -> extractExtraField(fieldsWithElement)),
                            extractMask(fieldsWithElement)));
                }
            }.doSwitch(0);
        } catch (final Exception e) {
            return empty();
        }
    }

    private static Integer nullableInteger(final Callable<Integer> callable) {
        return (Integer) nullableObject(callable);
    }

    private static Double nullableDouble(final Callable<Double> callable) {

        return (Double) nullableObject(callable);
    }

    private static String nullableString(final Callable<String> callable) {
        return (String) nullableObject(callable);
    }

    private static Boolean nullableBoolean(final Callable<Boolean> callable) {
        return (Boolean) nullableObject(callable);
    }

    private static JsonObject nullableJsonObject(final Callable<JsonObject> callable) {
        return (JsonObject) nullableObject(callable);
    }

    private static Object nullableObject(final Callable<?> callable) {
        try {
            return callable.call();
        } catch (Exception ignored) {
            return null;
        }
    }

    private static Integer extractVerticalRate(final String[] fields) {
        return (int) parseDouble(fields[16]);
    }

    private static Double extractTrack(final String[] fields) {
        return parseDouble(fields[13]);
    }

    private static Integer extractSquawk(final String[] fields) {
        return parseInt(fields[17]);
    }

    private static Boolean extractSpi(final String[] fields) {
        return parseFlag(fields[20]);
    }

    private static Boolean extractOnGround(final String[] fields) {
        return parseFlag(fields[21]);
    }

    private static int extractMask(final String[] fields) {
        if (fields.length >= 24) {
            return parseInt(fields[23]);
        }
        return 0;
    }

    private static JsonObject extractExtraField(final String[] fields) {
        if (fields.length == 23 || fields.length == 25) {
            JsonParser parser = new JsonParser();
            JsonObject jsonObject;
            try {
                jsonObject = parser.parse(fields[fields.length - 1]).getAsJsonObject();
            } catch (JsonParseException e) {
                System.out.println("Erreur de parsing JSON : " + e.getMessage());
                return null;
            }
            return jsonObject;
        }
        return null;

    }

    private static Double extractGroundSpeed(final String[] fields) {
        return parseDouble(fields[12]);
    }

    private static Boolean extractEmergency(final String[] fields) {
        return parseFlag(fields[19]);
    }

    private static Integer extractAltitude(final String[] fields) {
        try {
            return (int) parseDouble(fields[11]);
        } catch (final Exception ignored) {
            return null;
        }
    }

    private static Boolean extractAlert(final String[] fields) {
        return parseFlag(fields[18]);
    }

    private static Boolean parseFlag(final String flag) {
        return parseInt(flag) == 1;
    }

    private static int extractBstType(final String[] partLine) {
        return parseInt(partLine[1]);
    }

    private static int extractAircraftID(final String[] partLine) {
        return parseInt(partLine[3]);
    }

    private static int extractSessionID(final String[] partLine) {
        return parseInt(partLine[2]);
    }

    private static int extractFlightID(final String[] partLine) {
        return parseInt(partLine[5]);
    }

    private static String extractIcao(final String[] partLine) {
        return partLine[4];
    }

    private static long extractTimestampGenerated(final String[] partLine) throws ParseException {
        return strDateToTimestamp(partLine[6] + ',' + partLine[7]);
    }

    private static long extractTimestampLogged(final String[] partLine) throws ParseException {
        return strDateToTimestamp(partLine[8] + ',' + partLine[9]);
    }

    private static Double extractLatitude(final String[] partLine) {
        try {
            return parseDouble(partLine[14]);
        } catch (final Exception ignored) {
            return null;
        }
    }

    private static Double extractLongitude(final String[] partLine) {
        try {
            return parseDouble(partLine[15]);
        } catch (final Exception ignored) {
            return null;
        }
    }

    private static String extractCallSign(final String[] partLine) {
        return partLine[10];
    }

    public static long strDateToTimestamp(final String date) throws ParseException {
        final SimpleDateFormat dateFormat = new SimpleDateFormat(BST_DATE_PATTERN);
        dateFormat.setTimeZone(getTimeZone("UTC"));
        return dateFormat.parse(date).toInstant().toEpochMilli();
    }

    public static String timestampToStrDate(final long timestamp) {
        final SimpleDateFormat dateFormat = new SimpleDateFormat(BST_DATE_PATTERN);
        dateFormat.setTimeZone(getTimeZone("UTC"));
        return dateFormat.format(new Date(timestamp));
    }
}

File: packages/fdit-ts/src/fdit/alteration/core/basestation/BaseStationParser.ts
import BaseStationMessageFull from './message/BaseStationMessageFull'
import { Message } from '../engine/message'
import { JsonObject } from 'type-fest'
import { parse } from 'date-fns'
import { format } from 'date-fns-tz'

type Nullable<T> = T | null
type Callable<T> = () => T

function nullish<T>(callable: Callable<T>): Nullable<T> {
  try {
    return callable()
  } catch (ignored) {
    return null
  }
}

class BstMessageTypeSwitch<T> {
  private cases: { [key: string]: Callable<T> } = {}

  constructor(cases: { [key: string]: Callable<T> }) {
    this.cases = cases
  }

  doSwitch(key: string): T {
    if (this.cases.hasOwnProperty(key)) {
      return this.cases[key]()
    } else {
      throw new Error(`No case defined for key: ${key}`)
    }
  }
}

export class BaseStationParser {
  private static SPLIT_PATTERN = ',(?=(?:[^"]*"[^"]*")*[^"]*$)'
  private static SPLIT_JSON_PATTERN = ',\\{'
  public static BST_DATE_PATTERN = 'yyyy/MM/dd,HH:mm:ss.SSS'

  private constructor() {}

  public static createBstMessage(line: string): Message | null {
    const parts: string[] = line.split(BaseStationParser.SPLIT_JSON_PATTERN)
    const extraFieldsArePresent = parts.length > 1

    const fields: string[] = parts[0].split(BaseStationParser.SPLIT_PATTERN)
    const fieldsWithElement: (string[] | string)[] = extraFieldsArePresent
      ? [fields, `{${parts[1]}`]
      : fields

    try {
      const bstType: number =
        BaseStationParser.extractBstType(fieldsWithElement)
      const sessionID: number =
        BaseStationParser.extractSessionID(fieldsWithElement)
      const aircraftID: number =
        BaseStationParser.extractAircraftID(fieldsWithElement)
      const icao: string = BaseStationParser.extractIcao(fieldsWithElement)
      const flightID: number =
        BaseStationParser.extractFlightID(fieldsWithElement)
      const timestampGenerated: number =
        BaseStationParser.extractTimestampGenerated(fieldsWithElement)
      const timestampLogged: number =
        BaseStationParser.extractTimestampLogged(fieldsWithElement)

      return new BstMessageTypeSwitch<Message | null>({
        visitBstMessageFull: () => {
          return new BaseStationMessageFull(
            bstType,
            sessionID,
            aircraftID,
            icao,
            flightID,
            timestampGenerated,
            timestampLogged,
            BaseStationParser.nullableString(() =>
              BaseStationParser.extractCallSign(fieldsWithElement)
            ),
            BaseStationParser.nullableInteger(() =>
              BaseStationParser.extractAltitude(fieldsWithElement)
            ),
            BaseStationParser.nullableDouble(() =>
              BaseStationParser.extractGroundSpeed(fieldsWithElement)
            ),
            BaseStationParser.nullableDouble(() =>
              BaseStationParser.extractTrack(fieldsWithElement)
            ),
            BaseStationParser.nullableDouble(() =>
              BaseStationParser.extractLatitude(fieldsWithElement)
            ),
            BaseStationParser.nullableDouble(() =>
              BaseStationParser.extractLongitude(fieldsWithElement)
            ),
            BaseStationParser.nullableInteger(() =>
              BaseStationParser.extractVerticalRate(fieldsWithElement)
            ),
            BaseStationParser.nullableInteger(() =>
              BaseStationParser.extractSquawk(fieldsWithElement)
            ),
            BaseStationParser.nullableBoolean(() =>
              BaseStationParser.extractAlert(fieldsWithElement)
            ),
            BaseStationParser.nullableBoolean(() =>
              BaseStationParser.extractEmergency(fieldsWithElement)
            ),
            BaseStationParser.nullableBoolean(() =>
              BaseStationParser.extractSpi(fieldsWithElement)
            ),
            BaseStationParser.nullableBoolean(() =>
              BaseStationParser.extractOnGround(fieldsWithElement)
            ),
            BaseStationParser.nullableJsonObject(() =>
              BaseStationParser.extractExtraField(fieldsWithElement)
            )
            //, BaseStationParser.extractMask(fieldsWithElement)
          )
        },
      }).doSwitch('0')
    } catch (e) {
      return null
    }
  }

  private static nullableInteger(callable: Callable<number>): Nullable<number> {
    return nullish(callable)
  }

  private static nullableDouble(callable: Callable<number>): Nullable<number> {
    return nullish(callable)
  }

  private static nullableString(callable: Callable<string>): Nullable<string> {
    return nullish(callable)
  }

  private static nullableBoolean(
    callable: Callable<boolean>
  ): Nullable<boolean> {
    return nullish(callable)
  }

  private static nullableJsonObject(
    callable: Callable<JsonObject>
  ): Nullable<JsonObject> {
    return nullish(callable)
  }

  private static extractVerticalRate(fields: (string[] | string)[]): number {
    return parseFloat(<string>fields[16])
  }

  private static extractTrack(fields: (string[] | string)[]): number {
    return parseFloat(<string>fields[13])
  }

  private static extractSquawk(fields: (string[] | string)[]): number {
    return parseInt(<string>fields[17])
  }

  private static extractSpi(fields: (string[] | string)[]): boolean {
    return BaseStationParser.parseFlag(<string>fields[20])
  }

  private static extractOnGround(fields: (string[] | string)[]): boolean {
    return BaseStationParser.parseFlag(<string>fields[21])
  }

  private static extractMask(fields: string[]): number {
    return fields.length >= 24 ? parseInt(fields[23]) : 0
  }

  private static extractExtraField(
    fields: (string[] | string)[]
  ): JsonObject | null {
    if (fields.length === 23 || fields.length === 25) {
      return JSON.parse(<string>fields[fields.length - 1])
    }
    return null
  }

  private static extractGroundSpeed(fields: (string[] | string)[]): number {
    return parseFloat(<string>fields[12])
  }

  private static extractEmergency(fields: (string[] | string)[]): boolean {
    return BaseStationParser.parseFlag(<string>fields[19])
  }

  private static extractAltitude(fields: (string[] | string)[]): number | null {
    try {
      return parseInt(<string>fields[11])
    } catch (ignored) {
      return null
    }
  }

  private static extractAlert(fields: (string[] | string)[]): boolean {
    return BaseStationParser.parseFlag(<string>fields[18])
  }

  private static parseFlag(flag: string): boolean {
    return parseInt(flag) === 1
  }

  private static extractBstType(partLine: (string[] | string)[]): number {
    return parseInt(<string>partLine[1])
  }

  private static extractAircraftID(partLine: (string[] | string)[]): number {
    return parseInt(<string>partLine[3])
  }

  private static extractSessionID(partLine: (string[] | string)[]): number {
    return parseInt(<string>partLine[2])
  }

  private static extractFlightID(partLine: (string[] | string)[]): number {
    return parseInt(<string>partLine[5])
  }

  private static extractIcao(partLine: (string[] | string)[]): string {
    return <string>partLine[4]
  }

  private static extractTimestampGenerated(
    partLine: (string[] | string)[]
  ): number {
    return BaseStationParser.strDateToTimestamp(`${partLine[6]},${partLine[7]}`)
  }

  private static extractTimestampLogged(
    partLine: (string[] | string)[]
  ): number {
    return BaseStationParser.strDateToTimestamp(`${partLine[8]},${partLine[9]}`)
  }

  private static extractLatitude(
    partLine: (string[] | string)[]
  ): number | null {
    try {
      return parseFloat(<string>partLine[14])
    } catch (ignored) {
      return null
    }
  }

  private static extractLongitude(
    partLine: (string[] | string)[]
  ): number | null {
    try {
      return parseFloat(<string>partLine[15])
    } catch (ignored) {
      return null
    }
  }

  private static extractCallSign(partLine: (string[] | string)[]): string {
    return <string>partLine[10]
  }

  public static strDateToTimestamp(date: string): number {
    const dateFormat: Date = parse(
      date,
      BaseStationParser.BST_DATE_PATTERN,
      new Date(),
      { timeZone: 'UTC' }
    )
    return dateFormat.getTime()
  }

  public static timestampToStrDate(timestamp: number): string {
    return format(new Date(timestamp), BaseStationParser.BST_DATE_PATTERN, {
      timeZone: 'UTC',
    })
  }
}
